lib/pf/Switch/Ruckus/SmartZone.pm

```diff
@@ -49,38 +55,75 @@ sub parseExternalPortalRequest {
     return \%params;
 }
 
+=item deauthTechniques
 
-=item getAcceptForm
-
-Creates the form that should be given to the client device to trigger a reauthentication.
+Return the reference to the deauth technique or the default deauth technique.
 
 =cut
 
-sub getAcceptForm {
-    my ( $self, $mac, $destination_url, $portalSession ) = @_;
+sub deauthTechniques {
+    my ($self, $method) = @_;
     my $logger = $self->logger;
-    $logger->debug("Creating web release form");
-
-    my $client_ip = $portalSession->param("ecwp-original-param-uip");
-    my $controller_ip = $self->{_ip};
-
-    my $html_form = qq[
-        <form name="weblogin_form" action="http://$controller_ip:9997/SubscriberPortal/hotspotlogin" method="POST" style="display:none">
-          <input type="text" name="ip" value="$client_ip" />
-          <input type="text" name="username" value="$mac" />
-          <input type="text" name="password" value="$mac"/>
-          <input type="submit">
-        </form>
-
-        <script language="JavaScript" type="text/javascript">
-        window.setTimeout('document.weblogin_form.submit();', 1000);
-        </script>
-    ];
-
-    $logger->debug("Generated the following html form : ".$html_form);
-    return $html_form;
+    my $default = $SNMP::HTTP;
+    my %tech = (
+        $SNMP::HTTP => 'deauthenticateMacWebservices',
+    );
+
+    if (!defined($method) || !defined($tech{$method})) {
+        $method = $default;
+    }
+    return $method,$tech{$method};
 }
 
+sub deauthenticateMacWebservices {
+    my ($self, $mac) = @_;
+    my $logger = $self->logger;
+    my $controllerIp = $self->{_controllerIp} || $self->{_ip};
+    my $webservicesPassword = $self->{_wsPwd};
+    my $ucmac = uc $mac;
+    my $ip = pf::iplog::mac2ip($mac);
+    my $node_info = node_view($mac);
+    my $payload;
+    if($node_info->{status} eq "unreg" || violation_count_reevaluate_access($mac)) {
+        $payload = <<EOT;
+{
+ "Vendor": "ruckus",
+ "RequestPassword": "$webservicesPassword",
+ "APIVersion": "1.0",
+ "RequestCategory": "UserOnlineControl",
+ "RequestType": "Logout",
+ "UE-IP": "$ip",
+ "UE-MAC": "$ucmac"
+}
+EOT
+    }
+    else {
+        $payload = <<EOT;
+{
+ "Vendor": "ruckus",
+ "RequestPassword": "$webservicesPassword",
+ "APIVersion": "1.0",
+ "RequestCategory": "UserOnlineControl",
+ "RequestType": "Login",
+ "UE-IP": "$ip",
+ "UE-MAC": "$ucmac",
+ "UE-Proxy": "0",
+ "UE-Username": "$ucmac",
+ "UE-Password": "$ucmac"
+}
+EOT
```

Comment:
Use json_encode instead of building your json string manually.
Because if someone using a special character in the webservicesPassword like a '"' then it will fail.